bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdHandler.java

```diff
@@ -0,0 +1,633 @@
+import java.time.Duration;
+import java.time.Instant;
+import java.time.ZonedDateTime;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import javax.imageio.ImageIO;
+
+import org.apache.commons.lang.StringUtils;
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.api.ContentResponse;
+import org.eclipse.jetty.client.api.Request;
+import org.eclipse.jetty.http.HttpHeader;
+import org.eclipse.jetty.http.HttpMethod;
+import org.eclipse.jetty.http.HttpStatus;
+import org.eclipse.smarthome.core.common.ThreadPoolManager;
+import org.eclipse.smarthome.core.i18n.TimeZoneProvider;
+import org.eclipse.smarthome.core.library.types.DateTimeType;
+import org.eclipse.smarthome.core.library.types.DecimalType;
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.library.types.RawType;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.eclipse.smarthome.core.types.State;
+import org.eclipse.smarthome.core.types.UnDefType;
+import org.eclipse.smarthome.io.net.http.HttpRequestBuilder;
+import org.openhab.binding.doorbird.internal.listener.DoorbirdUdpListener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link DoorbirdHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Mark Hilbush - Initial contribution
+ */
+@NonNullByDefault
+public class DoorbirdHandler extends BaseThingHandler {
+    private static final long API_REQUEST_TIMEOUT = 3L;
+    private static final long MONTAGE_UPDATE_DELAY = 5L;
+
+    // Maximum number of doorbell and motion history images stored on Doorbird backend
+    private static final int MAX_HISTORY_IMAGES = 50;
+
+    private final Logger logger = LoggerFactory.getLogger(DoorbirdHandler.class);
+
+    // Get a dedicated threadpool for the listener thread
+    private final ScheduledExecutorService doorbirdScheduler = ThreadPoolManager.getScheduledPool("doorbirdHandler");
+    private @Nullable ScheduledFuture<?> listenerJob;
+    private DoorbirdUdpListener udpListener;
+
+    private @Nullable ScheduledFuture<?> imageRefreshJob;
+    private @Nullable ScheduledFuture<?> doorbellOffJob;
+    private @Nullable ScheduledFuture<?> motionOffJob;
+
+    private @Nullable String doorbirdId;
```

Comment:
No strong reason. Since I was testing for the presence of these config parameters in `initialize`, it seemed to make sense to save the values for use later. Would you prefer I just save `config`, then reference the parameters using that?

Reply:
Yes. If you do that, you can also simplify e.g. `startImageRefreshJob`, no need to use `getConfigAs` again. You can use `config.imageRefreshRate` then.

Reply:
Ok, then that's what I'll do. And, you're right, it does simplify a number of methods. 😉 

Since `getConfigAs` is defined as `@NonNull`, and `config` is set in the `initialize` method only, is this an appropriate use of `@NonNullByDefault({})` in the declaration?

```
private @NonNullByDefault({}) DoorbirdConfiguration config;
```